### ISSUE TITLE
wq.BasicTransport gets transposed, blocking save

### ISSUE THREAD

**Issue by Manangka** [Deltares/imod-python#294]
In GitLab by @jdelsman on Nov 17, 2022, 22:41

Not sure what has happened, but the BTN dataset somehow gets transposed to [layer, x, y]. I input [layer, y, x] dataarrays.

Not directly a solution to this specific problem, but a transpose in array_io.writing._save could stop some of the headache...
`if a.dims[-2:] != ("y", "x"):
    # try to transpose:
    _dims = [d for d in a.dims if d not in ("y", "x")]
    a = a.transpose(*_dims, "y", "x")
`
instead of:
`
    raise ValueError(
        'Last two dimensions of DataArray to save must be ("y", "x"), '
        f"found: {a.dims}"
    )`

**Comment by Manangka** [comment]
In GitLab by @Huite on Nov 21, 2022, 09:28

I've specifically added the check in 62d75e716a6e889d76eb7ff046e5f054771e25a4 because of #265.
It's very easy to accidentally transpose a DataArray, e.g. by using a time multiplier like this:

```python
da * time_multiplier  # (y, x) * (time,) -> (y, x, time)
```

instead of:

```python
time_multiplier * da  # (time,) * (y, x) -> (time, y, x)
```

This can then lead to all kinds of performance and memory issues... although maybe this can also/better be fixed by rechunking instead. I guess I'd have to run some benchmarks -- ideally with the original problem in #265.

But in any case, something is happening within the BTN save that broadcasts in the wrong order. 

There is some logic if a species is in the coordinates, but I don't think that should switch dim order around. And there doesn't seem to be much other logic.
Have you stepped through it with a debugger yet, which operation is causing the transpose/broadcast?

**Comment by Manangka** [comment]
In GitLab by @JoerivanEngelen on Mar 29, 2023, 18:34

FYI: I'm running into the same issue with running this workflow: https://gitlab.com/deltares/imod/california_model

Even when I force the input in the right order:

```python
imod.wq.BasicTransport(**ds.transpose("layer", "y", "x"))
```

**Comment by Manangka** [comment]
In GitLab by @JoerivanEngelen on Mar 30, 2023, 11:46

Turns out because of the ``thickness`` method of the BasicFlow package, where we do:

```python
        th = xr.concat(
            [
                self["top"]-self["bottom"].sel(layer=1),
                -1.0 * self["bottom"].diff("layer"),
            ],
            dim="layer",
        )
```

Apparently the behaviour in xarray changed in concat.

For reference, things I tried but did not fix the issue:

```python
        th = xr.concat(
            [
                -self["bottom"].sel(layer=1) + self["top"],
                -1.0 * self["bottom"].diff("layer"),
            ],
            dim="layer",
        )
```

The ``positions`` argument, also did not do what I expected (I thought you could use it to tweak the list order). 

So in the end, this is the easiest fix:

```python
    def thickness(self):
        """
        Computes layer thickness from top and bottom data.

        Returns
        -------
        thickness : xr.DataArray
        """
        th = xr.concat(
            [
                self["top"] - self["bottom"].sel(layer=1),
                -1.0 * self["bottom"].diff("layer"),
            ],
            dim="layer",
        )
        # Concat results in wrong dimension order, transpose to fix dimension order.
        return th.transpose("layer", "y", "x") 
```

**Comment by Manangka** [comment]
In GitLab by @JoerivanEngelen on Mar 30, 2023, 11:50

Fixed in: 9cc040091272a9332f8ce6f9fd1bcd8915953dc1
